### ISSUE TITLE
Enhancement: Rule to prefer type annotation over type assertion when possible

### ISSUE THREAD

**Issue by ark120202** [typescript-eslint/typescript-eslint#709]
```ts
type A = { a: string };
type B = A & { b: string };
const b: B = { a: 'a', b: 'b' };

const invalid = b as A;
const valid: A = b;

// Should this be handled by no-inferrable-types?
const invalid2: A = b as A;
```

Both `= b as A` and `: A = b` have the same result in this case, but `: A = b` looks cleaner to me and prevents potential bug if after refactoring `b` would stop extending `A`.

To implement this rule it has to make sure that the type of `b` is assignable to `A`. Type checker currently doesn't expose required API (there is a proposal: https://github.com/microsoft/TypeScript/issues/9879), however there is a [ts-simple-type](https://www.npmjs.com/package/ts-simple-type) module that reimplements typescript's logic.

**Comment by bradzacher** [comment]
Does `no-unnecessary-type-assertion` help here?

I could see this being added to `no-inferrable-types`

**Comment by ark120202** [comment]
> Does `no-unnecessary-type-assertion` help here?

No, it warns only if types are exactly equal

> If not I could see this being added to no-inferrable-types

I agree in the `: A = b as A` case, but not so sure about `= b as A`. The name of the rule suggests that the type is inferrable, but it actually requires a type hint, just in a different position. I could also see it in the `no-unnecessary-type-assertion` rule, but the `unnecessary` part there might be confusing as well.

**Comment by bradzacher** [comment]
I would probably lean toward this belonging in `no-object-literal-type-assertion`.
The rule should be renamed to `object-type-assertion`.

We'd have to use that `ts-simple-type` api (or the official api when/if it gets released)

**Comment by ark120202** [comment]
> object-type-assertion

I guess this rule also could be useful for non-objects:

```ts
type Color = 'red' | 'green' | 'blue';
const color = 'red';

const invalid = color as Color;
const valid: Color = color;
```

**Comment by bradzacher** [comment]
Thinking about it more, it make sense to roll it up with `no-angle-bracket-type-assertion` as well as `no-object-literal-type-assertion`.

New rule would be named `consistent-type-assertions`.

I.e. the rule would:
- ensure you use a consistent assertion style (`<cast>` vs `as cast`)
- ensure you use implicit vs explicit cast at the correct time.

**Comment by felixfbecker** [comment]
I honestly would love to a rule to forbid _all_ casts. In other languages you often need to cast, but in TypeScript you can almost always use type guards instead.

**Comment by bradzacher** [comment]
if you install the 2.0 canary, you'll be able to use the `consistent-type-assertions` rule, which has a `never` option!

**Comment by felixfbecker** [comment]
Awesome! But on second thought, I'm not sure if throwing all these different things into the same rule is good. Casting has _some_ use cases, and for a lot of these kind of rules, we often configure them with `warn` severity, so devs get tipped in their editor and can make a call, but won't have CI fail on them or have to spray `eslint:disable` everywhere. But if forbidding `<>` casting style and casts in general are the same rule, there would be no way to get these warnings, while also being strict about only allowing `as` casting style..

**Comment by bradzacher** [comment]
The problem is that I don't believe we can get enough information from the compiler to tell if an `value as X` is required vs `const variable: X`. 

Two rules would let you completely ban angle bracket assertions and also ban type assertions altogether, but that is a really weird way to configure your rules, right?

"don't use assertions but if you do break that rule, don't use angle brackets"

I would think that at the very least you should be checking closely enough in PRs to see that someone has done an eslint disable, otherwise someone could very easily just have two eslint disables and use an angle bracket assertion anyways.

i.e.

```ts
// eslint-disable-next-line @typescript-eslint/no-angle-bracket-assertions, @typescript-eslint/consistent-type-assertions
const x = <Foo>{ };

// is just as noticeable in a code review as

// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
const x = <Foo>{ };
```

**Comment by felixfbecker** [comment]
I see your point, but it's more like "warn against assertions in general, error on `<>` assertions" which doesn't sound that weird to me.

We definitely need to check in PRs for `eslint:disable` comments (although I think it's possible to forbid them with a CLI flag), but they are easy to spot and I think the bar is higher for devs to add those comments than to use inconsistent casting (that is usually not intentional, e.g. some code copied from StackOverflow). After all linting rules are all about saving code authors and reviewers time by giving feedback at the time of writing, so for me warnings are just as important of a tool as error-severity rules.

**Comment by ark120202** [comment]
I agree with points above. It's also a bit inconsistent now, since similar rules `prefer-namespace-keyword` and `no-namespace` are separated.

---

> The problem is that I don't believe we can get enough information from the compiler to tell if an `value as X` is required vs `const variable: X`.

I don't see any reason why it wouldn't be possible. I have made a small POC with ts-simple-type and it seems to work quite well. There are some comparison issues like https://github.com/runem/ts-simple-type/issues/32, but these can cause only false-negatives so it shouldn't be a huge problem.

**Comment by felixfbecker** [comment]
I just wanted to reiterate that the change to merge them together was not a good one in my eyes. I just upgraded us to 2.3.0, and can no longer configure ESLint like I used to before (and with TSLint) to error on inconsistent assertions while warning on object literal casts or casts in general. Since I want to hard-enforce a lint style, I now need to disable the warnings for object literal casts and casts completely, which reduces means our code is less safe and typescript-eslint is less useful to us. Warning against object literal casts or casts in general is not a _consistency_ issue, it is a _code safety_ issue. One could put all typescript-eslint rules into a single rule, but it would mean you can no longer configure them independently, and exactly that disadvantage is showing here with `consistent-type-assertions`. On the other hand, I see absolutely no _benefit_ to doing so. Please reconsider splitting out the options into their own rules 🙇

**Comment by bradzacher** [comment]
> similar rules prefer-namespace-keyword and no-namespace are separated

Definitely could merge them, I just hadn't done it.
It's rather low priority as those rules are pretty low value (most people don't ever reach for the namespace keyword anyways).

---

I don't see how there is more safety if you've got two rules.
If someone `eslint-disable-next-line` the rule, then they could still just as easily do it for two rules in one as I mentioned before.

So all you've done is traded one problem for another; both of which are code review problems:
- code review has to catch that there isn't an angle-bracket assertion when there is a disable comment.
- code review has to catch that there isn't an angle-bracket assertion when there is a disable comment, and has to catch that there aren't two disables at once, one of which is for the assertion style rule.

> I now need to disable the warnings for object literal casts and casts completely

But why? I don't understand. You can use the following config:
```
{
  assertionStyle: 'as',
  objectLiteralTypeAssertions: 'never',
}
```

Changing the object literal part of the rule from `never` to something more relaxed because you don't want people to use the wrong assertion when they disable the rule seems like the wrong approach.

---

> Since I want to hard-enforce a casting style

I might be misunderstanding, but it sounds like you're really asking for the following:

```ts
{
   "@typescript-eslint/assertion-style-if-you-really-have-to-assert": "as",
   "@typescript-eslint/object-literal-assertion-style": "never",
   "@typescript-eslint/non-object-literal-assertion-style": "never",
}
```

Which obviously involves 3 separate rules, one of which is old, one is new, and one is a variation on what currently exists.

Happy to accept PRs for this work, if you would like to do it.

---

This could also be solved by instead naming your files `.tsx`, which automatically hard bans all angle bracket assertions as an unavoidable, non-disableable _compile time error_.
I don't think one exists currently, but it would be very, very trivial to add a rule which checks filenames and reports if someone doesn't use `.tsx`:

```
module.exports = {
  create(context) {
    if (!context.getFilename().endsWith('.tsx') && !context.getFilename().endsWith('.d.ts')) {
      context.report({
        message: 'Files must use the extension \'.tsx\' or \'.d.ts\'',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 0  },
        },
      });
    }
    
    return {};
  },
};
```

---

Also, we're getting way, way, way off topic for this issue. If you would like to discuss this further - please raise a new issue.

**Comment by felixfbecker** [comment]
Because you cannot configure separate _severities_. `eslint-disable` has nothing to do with this, it is easy to ban `eslint-disable` comments. What I want to do (and what was possible before) is this:
```jsonc
{
   "@typescript-eslint/assertion-style": ["error", "as"],
   "@typescript-eslint/no-object-literal-assertions": "warn", // or no-assertions if existed
}
```

This way the assertion style is hard-enforced in CI, and you get _warnings_ in your editor for assertions that make you reconsider the cast and favor rewriting the code with a type annotation (but not fail CI). In a large team, these hints make code overall safer. Additionally, you can track the number of warnings in your codebase over time (post them as PR comments, show a graph in a dashboard, etc). You may also use the warning to roll out a rule gradually to eventually switch it to error once all instances have been resolved.

`eslint-disable` comments just get copied everywhere because people see them and assume they can just disable the rule everywhere, and they add more noise for patterns that are a code smell but still relatively common (and this is why it's great that the `warn` level exists).

This is now not possible anymore because the rules are in one and therefor share the same severity, and therefor I need to make the decision whether I either want no warnings for assertions but enforced `as` assertion style, or warnings for both, but `as` assertion style not hard enforced in CI.

Also note that the official ESLint rule naming guidelines say explicitly that rules that _disallow_ something (which the object literal assertion rule or the no-assertions rule clearly do, because you can't set it to "always") should be prefixed with `no-`: https://eslint.org/docs/developer-guide/working-with-rules#rule-naming-conventions

Renaming everything to TSX is not an alternative because there is some generic syntax that is impossible in TSX and editors also provide autocompletion (e.g. emmet shortcuts) that are unwanted in plain TS files.

I posted this here, instead of in a new issue, because I am interested in having the rule proposed in this issue, but your plan outlined above was to merge it into another rule, and that would expose the problem I am describing - so I am trying to propose to add it as a separate rule instead.

**Comment by JoshuaKGoldberg** [comment]
Btw, this really has been blocked on https://github.com/microsoft/TypeScript/issues/9879 -my old favorite!- the whole time.

**Comment by bradzacher** [comment]
This is also less of an issue now with the `satisfies` operator.
So we can just enforce the standard of using that over `as`!

**Comment by IchordeDionysos** [comment]
Is there a rule to prevent `as` and instead use `satisfies`?

**Comment by bradzacher** [comment]
No, because ultimately it's not always a 1:1 trade-off that you can make (same as not always being able to change from an `as` to a variable annotation).

You can ban `as` on your codebase if you like, see
https://typescript-eslint.io/linting/troubleshooting#how-can-i-ban-specific-language-feature

**Comment by Dremora** [comment]
@IchordeDionysos You can also consider [eslint-plugin-no-type-assertion](https://www.npmjs.com/package/eslint-plugin-no-type-assertion) (disclaimer: I'm the author). It allows to keep using `as const` and `as unknown` assertions, since these are safe.

**Comment by bradzacher** [comment]
We do have 
https://typescript-eslint.io/rules/consistent-type-assertions/

Which allows you to ban all assertions (that aren't const)

I forgot about it.
But no rule specifically to *suggest* using satisfies for specific cases.

**Comment by Dremora** [comment]
@bradzacher does this rule also exclude `as unknown` assertions? I'm happy to deprecate my plugin if its functionality can be fully covered by `typescript-eslint`.

**Comment by bradzacher** [comment]
@Dremora yes, it s you ban all assertions.
https://typescript-eslint.io/play#ts=4.9.3&sourceType=module&code=IwAghgziBMDcBQAKA3gX3FZ4D8AuEoqAlAgESkYgCuAdgNY0D2A7jQgNoC6lATgKZgAJoxoAbAJ4guCeABceVPpQDGIiLKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra0ipWgENkyRNHyUuqDAG1w2HFOgdokADTKVwKOMnTZTAMpEk6KE0QA3KZDABfXWAC6yl06A

**Comment by Dremora** [comment]
What I meant is that `as unknown` assertions are actually safe to have, so they don't need to be banned.

**Comment by JoshuaKGoldberg** [comment]
> `as unknown` assertions are actually safe to have

Most of the time. What about asserting a `never`-typed value to anything else, including `unknown`?

**Comment by Dremora** [comment]
> Most of the time. What about asserting a `never`-typed value to anything else, including `unknown`?

I can't think of a scenario of how one would acquire a `never` value during runtime, so asserting it to `unknown` should be allowed (although whatever code follows will simply never run). Perhaps I'm missing something?

**Comment by JoshuaKGoldberg** [comment]
One common pattern is to use an exhaustive set of `switch` cases to handle an API response, then put a console warning for an unexpected response type:

```ts
type ServerResponse = { type: 'a' } | { type: 'b' };

declare const response: ServerResponse;

switch (response.type) {
  case 'a': /* a logic */ break;
  case 'b': /* a logic */ break;
  default:
    console.warn("oh no:", response.type /* never */);
}
```

**Comment by Dremora** [comment]
@JoshuaKGoldberg that's an interesting scenario. If the code under the `default` branch does run, it implies the type definition was incomplete. In this case, it is safe to assert `response.type` as `unknown`, since it does have a value.

**Comment by JoshuaKGoldberg** [comment]
Now that v8 is released, per https://github.com/typescript-eslint/typescript-eslint/discussions/7936, we can mark this issue as unblocked! 🚀 

Except, I don't think we'd still want it in `consistent-type-assertions`.
* That rule has solidified as a stylistic one
* It doesn't require type information

...so unless there's an existing type-checked rule that covers this (I don't see one), I think we'd want a new dedicated rule? Thoughts @typescript-eslint/triage-team?

**Comment by bradzacher** [comment]
Personally - I think that banning "safe `as`" in favour of annotations and allowing "unsafe `as`" is not a great rule.
Instead I think allowing `as` in general and just banning unsafe `as`'s would be better (i.e. #7173).

I don't think there's a massive win to switching from `const x = { .. } as T;` to `const x: T = { .. };` is a massive win beyond styleistic concerns when you're guaranteeing that `as T` is safe. Esp considering there are cases where you'll need to do a safe `{ .. } as T` outside of a variable declaration and so that pattern will need to be in your codebase anways.

**Comment by JoshuaKGoldberg** [comment]
> win to switching from `const x = { .. } as T;` to `const x: T = { .. };` beyond stylistic concerns

+1, with the note that I think it is a win for stylistic concerns 😄. IMO this rule is valuable on its own as a pure stylistic one. Anecdotally, I see a lot of people new to TypeScript get confused over `: ` vs. `as`.

In other words, I think we'd benefit from adding a `prefer-type-annotation` rule that:

* Behavior: reports & autofixes any `Value as Type` to `Value: Type`
* When: the `Value` is, on its own, assignable to `Type` (the _"when possible"_ of this issue's title)

...but then the problem becomes: we don't have a good way with TypeScript APIs to know what the original ("on its own") type of `Type` would be. See also https://github.com/typescript-eslint/typescript-eslint/issues/9768#issuecomment-2316403066. So I think this is, unfortunately, not doable right now.
